core/src/main/scala/zio/config/ReadFunctions.scala

```diff
@@ -77,15 +85,37 @@ private[config] trait ReadFunctions {
     def loopZip[B, C](path: List[Step[K]], keys: List[K], cfg: Zip[K, V, B, C]): Res[(B, C)] =
       (loopAny(path, keys, cfg.left), loopAny(path, keys, cfg.right)) match {
         case (Right(leftV), Right(rightV)) => Right((leftV, rightV))
-        case (Left(leftE), Left(rightE))   => Left(AndErrors(leftE :: rightE :: Nil))
-        case (Left(leftE), _)              => Left(leftE)
-        case (_, Left(rightE))             => Left(rightE)
+        case (Left(leftE), Left(rightE)) =>
+          Left(AndErrors(leftE :: rightE :: Nil))
+        case (Left(leftE), _)  => Left(leftE)
+        case (_, Left(rightE)) => Left(rightE)
       }
 
     def loopXmapEither[B, C](path: List[Step[K]], keys: List[K], cfg: XmapEither[K, V, B, C]): Res[C] =
       loopAny(path, keys, cfg.config) match {
         case Left(error) => Left(error)
-        case Right(a)    => cfg.f(a).swap.map(ReadError.ConversionError(path.reverse, _)).swap
+        case Right(a) =>
+          cfg.f(a).swap.map(ReadError.ConversionError(path.reverse, _)).swap
+      }
+
+    def loopMap[B](path: List[Step[K]], keys: List[K], cfg: DynamicMap[K, V, B]): Res[Map[K, B]] =
+      cfg.source.getConfigValue(keys.reverse) match {
+        case PropertyTree.Leaf(_)     => formatError(path, "Leaf", "Record")
+        case PropertyTree.Sequence(_) => formatError(path, "Sequence", "Record")
+        case PropertyTree.Record(values) =>
+          val result: List[(K, Res[B])] = values.toList.zipWithIndex.map {
+            case ((k, tree), idx) =>
+              val source: ConfigSource[K, V] =
+                ConfigSource(cfg.source.sourceDescription, tree.getPath)
+
+              (k, loopAny(Step.Index(idx) :: path, Nil, cfg.config.updateSource(_ => source)))
+          }
+
+          seqMap2[K, ReadError[K], B, ReadError[K]]((index, k, error) => error.atKey(k).atIndex(index))(result.toMap).swap
```

Comment:
You don't need `atKey` and `atIndex` here (or anywhere else) since we have full path in `path` parameter.
Please, add test for error path if you think it's necessary.

Reply:
This is already in another branch as a follow up PR